File: university-service/src/test/java/ua/com/foxminded/courseproject/controllers/DayScheduleControllerTest.java
package ua.com.foxminded.courseproject.controllers;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.security.test.context.support.WithMockUser;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.util.MultiValueMap;
import ua.com.foxminded.courseproject.dto.DayScheduleDto;
import ua.com.foxminded.courseproject.dto.LessonDto;
import ua.com.foxminded.courseproject.service.DayScheduleService;

import java.time.LocalDate;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

import static org.mockito.Mockito.when;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@SpringBootTest
@AutoConfigureMockMvc
class DayScheduleControllerTest {
    @Autowired
    private MockMvc mockMvc;

    @MockBean
    private DayScheduleService service;

    Map<LocalDate, DayScheduleDto> dayScheduleDtoMap;

    @BeforeEach
    void setUp() {
        setDaySchedule();
    }

    @Test
    @WithMockUser(username = "admin", authorities = "ADMIN")
    void getTeacherSchedule_shouldReturnStartDayScheduleAndOk_WhenQueryStartDayOnly() throws Exception {
        LocalDate startDate = LocalDate.now();
        DayScheduleDto daySchedule = dayScheduleDtoMap.get(startDate);
        MultiValueMap<String, String> params = new LinkedMultiValueMap<>();
        String teacherId = UUID.randomUUID().toString();

        params.put("startdate", Collections.singletonList(startDate.toString()));

        when(service.getTeacherOneDaySchedule(startDate, UUID.fromString(teacherId))).thenReturn(daySchedule);

        mockMvc.perform(get("/teachers/{id}/schedule", teacherId).params(params))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.id").value(daySchedule.getId().toString()));
    }

    @Test
    @WithMockUser(username = "admin", authorities = "ADMIN")
    void getTeacherSchedule_shouldReturnDaysMapAndOk_WhenQueryEndDayBeforeStartDay() throws Exception {
        LocalDate startDate = LocalDate.now();
        LocalDate endDate = startDate.plusDays(1);
        MultiValueMap<String, String> params = new LinkedMultiValueMap<>();
        String teacherId = UUID.randomUUID().toString();

        params.put("startdate", Collections.singletonList(startDate.toString()));
        params.put("enddate", Collections.singletonList(endDate.toString()));

        when(service.getTeacherDaysSchedule(startDate, endDate, UUID.fromString(teacherId))).thenReturn(dayScheduleDtoMap);

        mockMvc.perform(get("/teachers/{id}/schedule", teacherId).params(params))
                .andDo(print())
                .andExpect(status().isOk())
                .andExpect(jsonPath("$." + startDate + ".id").
                        value(dayScheduleDtoMap.get(startDate).getId().toString()))
                .andExpect(jsonPath("$." + endDate + ".id").
                        value(dayScheduleDtoMap.get(endDate).getId().toString()));
    }

    @Test
    @WithMockUser(username = "admin", authorities = "ADMIN")
    void getTeacherSchedule_shouldReturnDayScheduleAndOk_WhenQueryEndDayBeforeStartDay() throws Exception {
        LocalDate startDate = LocalDate.now();
        LocalDate endDate = startDate.minusDays(1);
        DayScheduleDto daySchedule = dayScheduleDtoMap.get(startDate);
        MultiValueMap<String, String> params = new LinkedMultiValueMap<>();
        String teacherId = UUID.randomUUID().toString();

        params.put("startdate", Collections.singletonList(startDate.toString()));
        params.put("enddate", Collections.singletonList(endDate.toString()));

        when(service.getTeacherOneDaySchedule(startDate, UUID.fromString(teacherId))).thenReturn(daySchedule);

        mockMvc.perform(get("/teachers/{id}/schedule", teacherId).params(params))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.id").value(daySchedule.getId().toString()));
    }

    @Test
    @WithMockUser(username = "admin", authorities = "ADMIN")
    void getStudentSchedule_shouldReturnStartDayScheduleAndOk_WhenQueryStartDayOnly() throws Exception {
        LocalDate startDate = LocalDate.now();
        DayScheduleDto daySchedule = dayScheduleDtoMap.get(startDate);
        MultiValueMap<String, String> params = new LinkedMultiValueMap<>();
        String teacherId = UUID.randomUUID().toString();

        params.put("startdate", Collections.singletonList(startDate.toString()));

        when(service.getStudentOneDaySchedule(startDate, UUID.fromString(teacherId))).thenReturn(daySchedule);

        mockMvc.perform(get("/students/{id}/schedule", teacherId).params(params))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.id").value(daySchedule.getId().toString()));
    }

    @Test
    @WithMockUser(username = "admin", authorities = "ADMIN")
    void getStudentSchedule_shouldReturnDaysMapAndOk_WhenQueryEndDayBeforeStartDay() throws Exception {
        LocalDate startDate = LocalDate.now();
        LocalDate endDate = startDate.plusDays(1);
        MultiValueMap<String, String> params = new LinkedMultiValueMap<>();
        String studentId = UUID.randomUUID().toString();

        params.put("startdate", Collections.singletonList(startDate.toString()));
        params.put("enddate", Collections.singletonList(endDate.toString()));

        when(service.getStudentDaysSchedule(startDate, endDate, UUID.fromString(studentId))).thenReturn(dayScheduleDtoMap);

        mockMvc.perform(get("/students/{id}/schedule", studentId).params(params))
                .andDo(print())
                .andExpect(status().isOk())
                .andExpect(jsonPath("$." + startDate + ".id").
                        value(dayScheduleDtoMap.get(startDate).getId().toString()))
                .andExpect(jsonPath("$." + endDate + ".id").
                        value(dayScheduleDtoMap.get(endDate).getId().toString()));
    }

    @Test
    @WithMockUser(username = "admin", authorities = "ADMIN")
    void getStudentSchedule_shouldReturnDayScheduleAndOk_WhenQueryEndDayBeforeStartDay() throws Exception {
        LocalDate startDate = LocalDate.now();
        LocalDate endDate = startDate.minusDays(1);
        DayScheduleDto daySchedule = dayScheduleDtoMap.get(startDate);
        MultiValueMap<String, String> params = new LinkedMultiValueMap<>();
        String studentId = UUID.randomUUID().toString();

        params.put("startdate", Collections.singletonList(startDate.toString()));
        params.put("enddate", Collections.singletonList(endDate.toString()));

        when(service.getStudentOneDaySchedule(startDate, UUID.fromString(studentId))).thenReturn(daySchedule);

        mockMvc.perform(get("/students/{id}/schedule", studentId).params(params))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.id").value(daySchedule.getId().toString()));
    }

    private void setDaySchedule() {
        dayScheduleDtoMap = new HashMap<>();
        LessonDto lessonOne = new LessonDto();
        lessonOne.setId(UUID.randomUUID());
        LessonDto lessonTwo = new LessonDto();
        lessonTwo.setId(UUID.randomUUID());
        DayScheduleDto firstDay = new DayScheduleDto();
        firstDay.setId(UUID.randomUUID());
        firstDay.setDayNumber(1);
        firstDay.setLessons(Arrays.asList(lessonOne, lessonTwo));
        DayScheduleDto secondDay = new DayScheduleDto();
        secondDay.setId(UUID.randomUUID());
        secondDay.setDayNumber(2);
        secondDay.setLessons(Arrays.asList(lessonOne));
        dayScheduleDtoMap.put(LocalDate.now(), firstDay);
        dayScheduleDtoMap.put(LocalDate.now().plusDays(1), secondDay);
    }
}

File: university-service/src/test/java/ua/com/foxminded/courseproject/controllers/DayScheduleControllerTest.kt
package ua.com.foxminded.courseproject.controllers

import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.Mockito
import org.mockito.kotlin.any
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.reactive.AutoConfigureWebTestClient
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.security.test.context.support.WithMockUser
import org.springframework.test.web.reactive.server.WebTestClient
import org.springframework.util.LinkedMultiValueMap
import org.springframework.util.MultiValueMap
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import ua.com.foxminded.courseproject.dto.DayScheduleDto
import ua.com.foxminded.courseproject.dto.LessonDto
import ua.com.foxminded.courseproject.service.DayScheduleService
import java.time.LocalDate
import java.util.*

@SpringBootTest
@AutoConfigureWebTestClient
internal class DayScheduleControllerTest {
    @Autowired
    private lateinit var webTestClient: WebTestClient

    @MockBean
    private lateinit var service: DayScheduleService
    private lateinit var daySchedulePairs: Flux<Pair<LocalDate, DayScheduleDto?>>

    @BeforeEach
    fun setUp() {
        setDaySchedule()
    }

    @Throws(Exception::class)
    @WithMockUser(username = "admin", authorities = ["ADMIN"])
    @Test
    fun teacherSchedule_shouldReturnStartDayScheduleAndOk_WhenQueryStartDayOnly() {
        val startDate = LocalDate.now()
        val daySchedule = daySchedulePairs.filter { it.first == startDate }.blockFirst()?.second
        val params: MultiValueMap<String, String> = LinkedMultiValueMap()
        val teacherId = UUID.randomUUID().toString()
        params["startdate"] = listOf(startDate.toString())

        Mockito.`when`(service.getTeacherOneDaySchedule(any<LocalDate>(), any<UUID>()))
            .thenReturn(Mono.just(Pair(startDate, daySchedule)))

        webTestClient
            .get()
            .uri {
                it.path("/teachers/{id}/schedule")
                    .queryParams(params)
                    .build(teacherId)
            }
            .exchange()
            .expectStatus().isOk
            .expectBody()
            .jsonPath("$.[0].second.id").isEqualTo(daySchedule?.id.toString())
    }

    @Throws(Exception::class)
    @WithMockUser(username = "admin", authorities = ["ADMIN"])
    @Test
    fun teacherSchedule_shouldReturnDaysMapAndOk_WhenQueryEndDayBeforeStartDay() {
        val startDate = LocalDate.now()
        val endDate = startDate.plusDays(1)
        val params: MultiValueMap<String, String> = LinkedMultiValueMap()
        val teacherId = UUID.randomUUID().toString()
        val dayScheduleStart = daySchedulePairs.filter { it.first == startDate }.blockFirst()?.second
        val dayScheduleEnd = daySchedulePairs.filter { it.first == endDate }.blockFirst()?.second
        params["startdate"] = listOf(startDate.toString())
        params["enddate"] = listOf(endDate.toString())

        Mockito.`when`(service.getTeacherDaysSchedule(startDate, endDate, UUID.fromString(teacherId)))
            .thenReturn(daySchedulePairs)

        webTestClient
            .get()
            .uri {
                it.path("/teachers/{id}/schedule")
                    .queryParams(params)
                    .build(teacherId)
            }
            .exchange()
            .expectStatus().isOk
            .expectBody()
            .jsonPath("$.[0].second.id").isEqualTo(dayScheduleStart?.id.toString())
            .jsonPath("$.[1].second.id").isEqualTo(dayScheduleEnd?.id.toString())
    }

    @Throws(Exception::class)
    @WithMockUser(username = "admin", authorities = ["ADMIN"])
    @Test
    fun teacherSchedule_shouldReturnDayScheduleAndOk_WhenQueryEndDayBeforeStartDay() {
        val startDate = LocalDate.now()
        val endDate = startDate.minusDays(1)
        val daySchedule = daySchedulePairs.filter { it.first == startDate }.blockFirst()?.second
        val params: MultiValueMap<String, String> = LinkedMultiValueMap()
        val teacherId = UUID.randomUUID().toString()
        params["startdate"] = listOf(startDate.toString())
        params["enddate"] = listOf(endDate.toString())

        Mockito.`when`(service.getTeacherOneDaySchedule(startDate, UUID.fromString(teacherId)))
            .thenReturn(Mono.just(Pair(startDate, daySchedule)))

        webTestClient
            .get()
            .uri {
                it.path("/teachers/{id}/schedule")
                    .queryParams(params)
                    .build(teacherId)
            }
            .exchange()
            .expectStatus().isOk
            .expectBody()
            .jsonPath("$.[0].second.id").isEqualTo(daySchedule?.id.toString())
    }

    @Throws(Exception::class)
    @WithMockUser(username = "admin", authorities = ["ADMIN"])
    @Test
    fun studentSchedule_shouldReturnStartDayScheduleAndOk_WhenQueryStartDayOnly() {
        val startDate = LocalDate.now()
        val daySchedule = daySchedulePairs.filter { it.first == startDate }.blockFirst()?.second
        val params: MultiValueMap<String, String> = LinkedMultiValueMap()
        val teacherId = UUID.randomUUID().toString()
        params["startdate"] = listOf(startDate.toString())

        Mockito.`when`(service.getStudentOneDaySchedule(startDate, UUID.fromString(teacherId)))
            .thenReturn(Mono.just(Pair(startDate, daySchedule)))

        webTestClient
            .get()
            .uri { it
                .path("/students/{id}/schedule")
                .queryParams(params)
                .build(teacherId)
            }
            .exchange()
            .expectStatus().isOk
            .expectBody()
            .jsonPath("$.[0].second.id").isEqualTo(daySchedule?.id.toString())
    }

    @Throws(Exception::class)
    @WithMockUser(username = "admin", authorities = ["ADMIN"])
    @Test
    fun studentSchedule_shouldReturnDaysMapAndOk_WhenQueryEndDayBeforeStartDay() {
        val startDate = LocalDate.now()
        val endDate = startDate.plusDays(1)
        val params: MultiValueMap<String, String> = LinkedMultiValueMap()
        val studentId = UUID.randomUUID().toString()
        val dayScheduleStart = daySchedulePairs.filter { it.first == startDate }.blockFirst()?.second
        val dayScheduleEnd = daySchedulePairs.filter { it.first == endDate }.blockFirst()?.second
        params["startdate"] = listOf(startDate.toString())
        params["enddate"] = listOf(endDate.toString())

        Mockito.`when`(service.getStudentDaysSchedule(startDate, endDate, UUID.fromString(studentId)))
            .thenReturn(daySchedulePairs)

        webTestClient
            .get()
            .uri { it
                .path("/students/{id}/schedule")
                .queryParams(params)
                .build(studentId)
            }
            .exchange()
            .expectStatus().isOk
            .expectBody()
            .jsonPath("$.[0].second.id").isEqualTo(dayScheduleStart?.id.toString())
            .jsonPath("$.[1].second.id").isEqualTo(dayScheduleEnd?.id.toString())
    }

    @Throws(Exception::class)
    @WithMockUser(username = "admin", authorities = ["ADMIN"])
    @Test
    fun studentSchedule_shouldReturnDayScheduleAndOk_WhenQueryEndDayBeforeStartDay() {
        val startDate = LocalDate.now()
        val endDate = startDate.minusDays(1)
        val daySchedule = daySchedulePairs.filter { it.first == startDate }.blockFirst()?.second
        val params: MultiValueMap<String, String> = LinkedMultiValueMap()
        val studentId = UUID.randomUUID().toString()
        params["startdate"] = listOf(startDate.toString())
        params["enddate"] = listOf(endDate.toString())

        Mockito.`when`(service.getStudentOneDaySchedule(startDate, UUID.fromString(studentId)))
            .thenReturn(Mono.just(Pair(startDate, daySchedule)))

        webTestClient
            .get()
            .uri {
                it
                    .path("/students/{id}/schedule")
                    .queryParams(params)
                    .build(studentId)
            }
            .exchange()
            .expectStatus().isOk
            .expectBody().jsonPath("$.[0].second.id").isEqualTo(daySchedule?.id.toString())
    }

    private fun setDaySchedule() {
        val lessonOne = LessonDto()
        lessonOne.id = UUID.randomUUID()
        val lessonTwo = LessonDto()
        lessonTwo.id = UUID.randomUUID()
        val firstDay = DayScheduleDto()
        firstDay.id = UUID.randomUUID()
        firstDay.dayNumber = 1
        firstDay.lessons = Arrays.asList(lessonOne, lessonTwo)
        val secondDay = DayScheduleDto()
        secondDay.id = UUID.randomUUID()
        secondDay.dayNumber = 2
        secondDay.lessons = Arrays.asList(lessonOne)
        daySchedulePairs = Flux.just(Pair(LocalDate.now(), firstDay), Pair(LocalDate.now().plusDays(1), secondDay))
    }
}